framework.cpp

```diff
@@ -0,0 +1,403 @@
+            out.push_back(in[i->first]);
+        }
+        return out;
+    }
+
+
+    /* // STL and Blaze versions apart, not needed in this case
+    template <typename set_t_l>
+    typename std::enable_if <
+     std::is_same<typename determine_val_t<set_t_l>::type, typename Metric::Rec_T>::value
+       && determine_container_type<typename determine_val_t<set_t_l>::type>::code == 1, // STL
+     set_t_l
+    >::type
+    sort_by_distance(set_t_l in, typename Metric::Rec_T center) {
+
+        set_t_l out;
+
+        auto cmp = [](std::pair<size_t, double> const & a, std::pair<size_t, double> const & b)
+        {
+            return a.second != b.second ? a.second < b.second : a.first < b.first;
+        };
+
+        std::vector<std::pair<size_t, double>> distances;
+        for (size_t i = 0; i<in.size(); i++) {
+            distances.push_back(std::make_pair(i, metric(in[i], center)));
+        }
+        sort(distances.begin(), distances.end(), cmp);
+        for (std::vector<std::pair<size_t, double>>::iterator i = distances.begin(); i != distances.end(); i++) {
+            out.push_back(in[i->first]);
+        }
+        return out;
+    }
+
+
+    template <typename set_t_l>
+    typename std::enable_if <
+     std::is_same<typename determine_val_t<set_t_l>::type, typename Metric::Rec_T>::value
+       && determine_container_type<typename determine_val_t<set_t_l>::type>::code == 2, // Blaze
+     set_t_l
+    >::type
+    sort_by_distance(set_t_l in, typename Metric::Rec_T center) { // only STP version now
+
+        set_t_l out (in.size());
+
+        auto cmp = [](std::pair<size_t, double> const & a, std::pair<size_t, double> const & b)
+        {
+            return a.second != b.second ? a.second < b.second : a.first < b.first;
+        };
+
+        std::vector<std::pair<size_t, double>> distances;
+        for (size_t i = 0; i<in.size(); i++) {
+            distances.push_back(std::make_pair(i, metric(in[i], center)));
+        }
+        sort(distances.begin(), distances.end(), cmp);
+        for (size_t i = 0; i<distances.size(); i++) {
+            out[i] = in[distances[i].first];
+        }
+        return out;
+    }
+    // */
+
+
+protected:
+
+    Metric metric;
+};
+
+
+
+
+// --------
+// derived classes
+
+
+template <typename rec_t>
+class EuclideanNoIter : public MetricBase<rec_t> {
+
+public:
+
+    double operator()(const rec_t & a, const rec_t & b) const  { // siuts both Blaze and STL
+        double sum = 0;
+        if (a.size() != b.size()) {
+            double dnan = std::nan("Input container sizes do not match");
+            return dnan;
+        }
+        for (size_t i = 0; i < a.size(); ++i) {
+            sum += (a[i] - b[i]) * (a[i] - b[i]);
+        }
+        return std::sqrt(sum);
+    }
+
+};
+
+
+
+template <typename rec_t>
+class EuclideanTypeSpecific : public MetricBase<rec_t> {
```

Comment:
while reading this I caught myself multiple time on confusing `rec_t` with `rec_t_l`, `set_t_l` and `set_t`, :)
is there any style guide you are following here?

Reply:
yes, this is a bit annoying. I'm not sure about naming at all
set_t, val_t, rec_t, met_t
Set, Value, Record, Metric
SetType, ValueType, RecordType, MetricType

Reply:
in the current naming, _l stands for local template parameter, capital letters appear in names of public fields.. Maybe it'd be better to use capitals for template parameters

Reply:
also, my vote is:
SetType, ValueType, RecordType, MetricType for class template arguments
set_type, value_type, record_type, metric_type for public class members (note that it's consistent with naming used in STL)
and maybe something like SetTypeL for method templates (local, only inner used, not exposed to client code)

Reply:
updated "Blaze-style" naming proposition:
ST, VT, RT, MT for set, value, record and metric return type template arguments respectively,
S, V, R, M for public class members
and maybe S_ for inner method templates?

Reply:
what is short template parameter naming for for metric type? MT?